services/debugger/destination/eventDeliveryStatusUploader.go

```diff
@@ -35,23 +35,43 @@ var uploader debugger.UploaderI
 var (
 	configBackendURL                  string
 	disableEventDeliveryStatusUploads bool
+	eventsDeliveryCacheMap            map[string][]*DeliveryStatusT
+	eventsCacheMapLock                sync.RWMutex
+	maxEventsCacheSize  int
 )
 
 var pkgLogger logger.LoggerI
 
 func Init() {
 	loadConfig()
 	pkgLogger = logger.NewLogger().Child("debugger").Child("destination")
+	eventsDeliveryCacheMap = make(map[string][]*DeliveryStatusT)
 }
 
 func loadConfig() {
 	configBackendURL = config.GetEnv("CONFIG_BACKEND_URL", "https://api.rudderlabs.com")
 	config.RegisterBoolConfigVariable(false, &disableEventDeliveryStatusUploads, true, "DestinationDebugger.disableEventDeliveryStatusUploads")
+	config.RegisterIntConfigVariable(2, &maxEventsCacheSize, true, 1, "DestinationDebugger.maxEventsCacheSize")
 }
 
 type EventDeliveryStatusUploader struct {
 }
 
+func populateEventsCache(cache map[string][]*DeliveryStatusT, destinationID string, deliveryStatus *DeliveryStatusT) {
+	eventsCacheMapLock.Lock()
+	defer eventsCacheMapLock.Unlock()
+
+	if _, ok := cache[destinationID]; !ok {
+		cache[destinationID] = make([]*DeliveryStatusT, 0, maxEventsCacheSize)
+	}
+	eventsCache := cache[destinationID]
+	eventsCache = append(eventsCache, deliveryStatus)
+	if len(eventsCache) > maxEventsCacheSize {
+		eventsCache = eventsCache[len(eventsCache)-maxEventsCacheSize:]
+	}
+	cache[destinationID] = eventsCache
```

Comment:
An alternate way to implement this would be a ring buffer:
https://github.com/smallnest/ringbuffer
https://pkg.go.dev/container/ring

It gives performance benefits (no memory allocation) and you could avoid Locks at some extend.

However, it would be more complex and I am not sure if our use case here justifies this complexity.

Reply:
See if we can avoid allocation of new memory every time we add element in the cache.